File: 백준/Silver/9012. 괄호/괄호.py
import sys

N = int(sys.stdin.readline().strip())

for i in range(N):
  stack = []
  flag = True
  temp_str = sys.stdin.readline().strip()
  for j in range(len(temp_str)):
    if temp_str[j] =='(': 
      stack.append('(')
    elif temp_str[j] == ')':
      if len(stack) == 0:
        flag = False
        break
      else:
        stack.pop(-1)
  
  if len(stack)!=0: flag=False    
  if flag==False:
    print("NO")
  else:
    print("YES")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        String n = bf.readLine();
        int N = Integer.parseInt(n);

        for(int q=0; q<N; q++){
            Stack<Integer> s = new Stack<>();
            char[] temp = bf.readLine().toCharArray();
            boolean check = false;
            for(int i=0; i<temp.length; i++){
                if(temp[i] == '('){
                    s.push(0);
                }else{
                    if(s.isEmpty()){
                        System.out.println("NO");
                        check = true;
                        break;
                    }else{
                        s.pop();
                    }
                }
            }
            if(!check){
                if(s.isEmpty()){
                    System.out.println("YES");
                }else{
                    System.out.println("NO");
                }
            }
        }
    }
}